Detectors/PHOS/calib/src/PHOSEnergyCalibDevice.cxx

```diff
@@ -85,8 +100,76 @@ void PHOSEnergyCalibDevice::endOfStream(o2::framework::EndOfStreamContext& ec)
   if (mPostHistos) {
     postHistosCCDB(ec);
   }
+  writeOutFile();
+}
+
+void PHOSEnergyCalibDevice::stop()
+{
+  // DDS stop method: simply write the latest tree
+  LOG(info) << "STOP: writing output digits file";
+  writeOutFile();
+}
+void PHOSEnergyCalibDevice::fillOutputTree()
+{
+
+  if (mOutputDigits.size() == 0) { // nothing to fill
+    return;
+  }
+
+  LOG(info) << "Filling tree with " << mOutputDigits.size() << " digits";
+  if (!mFileOut) { // create file and tree
+    mFileName = mOutputDir + fmt::format("PHOS_CalibDigits_{}.root", mRunNumber);
+    LOG(info) << "Creating new tree for PHOS calib digits, output file=" << mFileName;
+    mFileOut = std::make_unique<TFile>(mFileName.c_str(), "recreate");
+    mTreeOut = std::make_unique<TTree>("phosCalibDig", "O2 PHOS calib tree");
+    mFileMetaData = std::make_unique<o2::dataformats::FileMetaData>();
+  }
+  auto* br = mTreeOut->GetBranch("PHOSCalib");
+  auto* pptr = &mOutputDigits;
+  if (br) {
+    LOG(info) << " Using existing branch";
+    br->SetAddress(&pptr);
+  } else {
+    LOG(info) << " Create new branch";
+    br = mTreeOut->Branch("PHOSCalib", &pptr);
+  }
+  int res = br->Fill();
+  br->ResetAddress();
 }
 
+void PHOSEnergyCalibDevice::writeOutFile()
+{
+  // write collected vector and metadata
+  if (!mTreeOut) { // nothing to write,
+    return;
+  }
+  LOG(info) << "Writing calibration digits";
+  mFileOut->cd();
+  int nbits = mTreeOut->Write();
+  LOG(info) << "Wrote " << nbits << " bits";
+  mTreeOut.reset();
+  mFileOut->Close();
+  mFileOut.reset();
+
+  // write metaFile data
+  mFileMetaData->fillFileData(mFileName);
+  mFileMetaData->run = mRunNumber;
+  mFileMetaData->LHCPeriod = mLHCPeriod;
+  mFileMetaData->type = "calib";
+  mFileMetaData->priority = "high";
+  auto metaFileNameTmp = mFileName + ".tmp";
+  auto metaFileName = mFileName + ".done";
+  try {
+    std::ofstream metaFileOut(metaFileNameTmp);
+    metaFileOut << *mFileMetaData.get();
+    metaFileOut.close();
+    std::filesystem::rename(metaFileNameTmp, metaFileName);
+  } catch (std::exception const& e) {
+    LOG(error) << "Failed to store PHOS meta data file " << metaFileName << ", reason: " << e.what();
+  }
+  LOG(info) << "Stored metadate file " << mFileName << ".done";
+  mFileMetaData.reset();
```

Comment:
metadate file cannot be in the same directory as the root file it describes, you should have a separate directory option for it, see how it is done in the CTFWriterSpec.cxx

Reply:
Both issues fixed. I moved options to PHOSEnergyCalibDevice class, to reduce the number of parameters to be transferred.

Reply:
Not really: your `mFileName` is defined as `mOutputDir + fmt::format("PHOS_CalibDigits_{}.root", mRunNumber);`
https://github.com/AliceO2Group/AliceO2/blob/c08bfdf79038d1147044b7e6c3cba3257db9e202/Detectors/PHOS/calib/src/PHOSEnergyCalibDevice.cxx#L143
Here you are creating a metafile with the name `<mFileName>.tmp` then renaming it to  `<mFileName>.done`, the `mMetaFileDir` is unused. Instead, you should 1st create e.g,
`mMetaFileDir + fmt::format("PHOS_CalibDigits_{}.root.tmp", mRunNumber);` then, once closed, rename it to
`mMetaFileDir + fmt::format("PHOS_CalibDigits_{}.root.done", mRunNumber);`